### DISCUSSION TITLE
[Feature] Improved caching on mobile app

### DISCUSSION THREAD

**Discussion by justin13888** [immich-app/immich#9706] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

When using the immich Mobile app, there seems to be multiple possible improvements to make large photo albums for skimmable. In particular, a few features that I believe would be valuable include:

1. Option to cache low resolution thumbnails and even original assets on device
2. Viewing photo album since last connection to server (i.e. old album is visible, new photos on device since last opened is visible immediately before attempting to sync with server)

In general, it's important for the app to feel as close to photos stored on device. Could assume users has slow or no Internet in worse case and find improvements to the load speed.

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by AngelaDMerkel** [comment]
In general I find the application much less responsive then the web view. Loading is virtually instant for me on web but takes a relatively long time on the application.

**Comment by scoop96** [comment]
I would love to see the cache in the phone app. I do have slow connection at home server, and would be nice to be able to find pictures quicker when im outside of home.

**Comment by cheeregowh** [comment]
I’m experiencing a similar issue. I often access external networks, and the transfer speed is quite slow. Every time I open the app after some time, the previously downloaded preview images are automatically cleared. As a result, my device has to reconnect to the NAS network to re-download the preview images, causing delays in loading them, which negatively impacts the user experience.

What could be the reason for this? Doesn't the app have a caching mechanism? I noticed that within a few hours of closing the app, the storage usage of the app doesn’t change, and the preview images can be quickly displayed when I reopen it. However, after a longer period, such as the next day, I find that the thumbnails have been cleared, and the app’s storage usage decreases, indicating that the cached data was automatically cleared.

**Comment by radh21301** [comment]
I have terrible speeds when accessing my server. I have my server in a different country, and the speeds are terrible. 

I would absolutely love to cache all the images (thumbnails and previews) in the app. They need much less space than the original, so I would assume quite a few people would have enough storage for these in their mobile devices.

Right now, I mostly just avoid using Immich altogether when I am away from home. The experience is quite bad when the internet is slow.

**Reply by AngelaDMerkel** [reply]
This shouldn't make too much of a difference in how fast things load. It would have more of an impact over latency, I would think

**Reply by radh21301** [reply]
You might be right.

I was referring to the scenario where I scroll to a certain point in the timeline, and wait about 10 seconds for the thumbnails to load.

And once you click on a thumbnail, the view initially opens the same thumbnail which is blurry. Then after about 2 seconds a slightly better version appears (which I assume is the preview).

**Comment by kfstorm** [comment]
I miss the functionality that Google Photos downloads all thumbnails automatically to my phone to make them available locally and offline. It's very unproductive to scroll and wait thumbnails to load when you try to find a photo but don't know the exact date of taken.

**Comment by advert665** [comment]
The music app Symphonium has a rolling cache where you can choose how big you want the cache to be. It will store newly streamed songs until it is full, and then start removing those that you haven't listened to for a while. It would be really great if this could be implemented on the immich mobile app. 

I would personally design it to prioritise thumbnails and then the full-size images. So thumbnails are always prioritised (if thumbnails fill the cache then the cache only includes thumbnails and removes the oldest viewed thumbnails), then full size images (if full size images fill the cache then it removes the oldest viewed full size images but leaves the thumbnails).

![WhatsApp Image 2025-05-14 at 14 54 52_4458e14d](https://github.com/user-attachments/assets/d21f6446-f15c-4d38-a965-dac225c81ef2)

**Comment by uri200** [comment]
Adding a concrete use case for point 1 (cache low-resolution thumbnails on device), plus a sketch of what I think the minimum viable version looks like.

**Use case: the server isn't always on.** My Immich server is a home machine that I power off when I'm not using it, to save energy. It's reachable remotely over HTTPS through a reverse proxy — but only while it's actually running. When it's off, the mobile app is effectively unusable: no browsing at all. I can't even look at the library to decide whether it's worth powering the server on.

To be clear about scope, this isn't about my phone's own camera roll (that's #4466). It's the *library* — the assets that only live on the server, which is most of them.

**What would fix it:** an opt-in setting to download and permanently keep the thumbnail (and optionally preview) version of every asset in the library.

- **Timeline works fully offline** from the local thumbnails — scroll the whole library, jump by date, open albums.
- **Tapping a photo:** if the server is reachable, load the full-quality original exactly as today. If not, show the cached lower-quality version, with a small badge so it's unambiguous that you're looking at the offline preview and not the real thing.
- **Background sync** whenever the server is reachable, so new uploads get their local copies without any manual action.
- **Show the cost up front** before enabling — e.g. "12,000 assets ≈ 800 MB" — and provide a "clear cache" button.

**Why thumbnails specifically, and why the whole library.** Thumbnails are small enough that the entire library fits on a phone. That's what makes this different from the per-album offline requests over in #7627: I don't want to pre-pick a subset, I want to be able to *find* any photo I've ever taken while offline, and then fetch the original once I'm back online. Originals stay server-side, where the storage cost belongs.

It also addresses the eviction complaints in this thread and in #15334 from a different angle. The ask isn't a bigger LRU cache — it's a cache that is explicitly **not** evictable, because the user opted in and was shown exactly what it costs.

Related threads covering overlapping pieces of this: #4466, #7627, #21017, #28496, #13103.
